// @multi-ai/orchestrator/dist/orchestrator/src/ProviderWorker.js
import { BrowserManager, waitForNetworkSettled } from "@multi-ai/browser-runner";
export class ProviderWorker {
    provider;
    browserManager = new BrowserManager();
    refreshRetryProviders = new Set(["gemini", "kimi"]);
    constructor(provider) {
        this.provider = provider;
    }
    async run(ctx, prompt) {
        const startedAt = Date.now();
        try {
            const runtimeContext = {
                ...ctx,
                session: await this.browserManager.getSession(ctx.providerId, ctx.visible)
            };
            return await this.runWithRefreshRetry(runtimeContext, prompt, startedAt);
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            const { status, errorCode, errorMessage } = this.classifyFailure(ctx.providerId, message);
            return {
                providerId: this.provider.id,
                status,
                errorCode,
                errorMessage,
                elapsedMs: Date.now() - startedAt
            };
        }
    }
    async runWithRefreshRetry(ctx, prompt, startedAt) {
        let initialError = null;
        for (let attempt = 0; attempt < 2; attempt += 1) {
            try {
                if (attempt > 0) {
                    await this.prepareRefreshRetry(ctx);
                }
                const elapsed = Date.now() - startedAt;
                const remainingMs = Math.max(10000, ctx.timeoutMs - elapsed);
                const currentCtx = { ...ctx, timeoutMs: remainingMs };
                await this.ensureLoggedIn(currentCtx);
                return await this.executePrompt(currentCtx, prompt, startedAt);
            }
            catch (error) {
                const normalizedError = error instanceof Error ? error : new Error(String(error));
                if (attempt === 0 &&
                    this.shouldRetryAfterRefresh(ctx.providerId, normalizedError.message)) {
                    initialError = normalizedError;
                    continue;
                }
                if (initialError) {
                    throw new Error(`${this.provider.name} refresh retry failed. First error: ${initialError.message}. Final error: ${normalizedError.message}`, { cause: error });
                }
                throw normalizedError;
            }
        }
        throw initialError ?? new Error(`${this.provider.name} failed without a retryable error`);
    }
    async ensureLoggedIn(ctx) {
        const t0 = Date.now();
        const loggedIn = await this.provider.checkLogin(ctx);
        console.log(`[${this.provider.name}] checkLogin: ${loggedIn ? '✅' : '❌'} (${Date.now() - t0}ms)`);
        if (loggedIn) {
            return;
        }
        await this.provider.openLogin(ctx);
        await this.provider.waitForManualLogin(ctx);
    }
    async executePrompt(ctx, prompt, startedAt) {
        const t = (label, since) => console.log(`[${this.provider.name}] ${label}: ${Date.now() - since}ms`);
        const t0 = Date.now();
        await this.provider.openHome(ctx);
        t('→ openHome', t0);
        const t1 = Date.now();
        await this.provider.ask(ctx, prompt);
        t('→ ask (input+submit)', t1);
        const t2 = Date.now();
        await this.provider.waitForAnswerComplete(ctx);
        t('→ waitForAnswerComplete', t2);
        const t3 = Date.now();
        const answer = await this.provider.extractAnswer(ctx, prompt);
        t('→ extractAnswer', t3);
        console.log(`[${this.provider.name}] 总耗时: ${Date.now() - startedAt}ms`);
        return {
            providerId: this.provider.id,
            status: "completed",
            answer,
            elapsedMs: Date.now() - startedAt
        };
    }
    async prepareRefreshRetry(ctx) {
        ctx.session = await this.browserManager.getSession(ctx.providerId, ctx.visible);
        const page = ctx.session.page;
        if (page.isClosed() || page.url() === "about:blank") {
            await this.provider.openHome(ctx);
            return;
        }
        try {
            await page.reload({ waitUntil: "domcontentloaded" });
            await waitForNetworkSettled(page);
        }
        catch {
            await this.provider.openHome(ctx);
        }
    }
    shouldRetryAfterRefresh(providerId, message) {
        if (!this.refreshRetryProviders.has(providerId)) {
            return false;
        }
        const normalized = message.toLowerCase();
        if (normalized.includes("manual verification") ||
            normalized.includes("challenge") ||
            normalized.includes("captcha") ||
            normalized.includes("verification") ||
            normalized.includes("requires login") ||
            normalized.includes("manual login")) {
            return false;
        }
        return true;
    }
    classifyFailure(providerId, message) {
        const normalized = message.toLowerCase();
        if (normalized.includes("manual verification") ||
            normalized.includes("challenge") ||
            normalized.includes("captcha") ||
            normalized.includes("verification")) {
            return {
                status: "failed",
                errorCode: "MANUAL_VERIFICATION_REQUIRED",
                errorMessage: message
            };
        }
        if (normalized.includes("timed out")) {
            return {
                status: "timeout",
                errorCode: "PROVIDER_TIMEOUT",
                errorMessage: message
            };
        }
        if (normalized.includes("requires login") || normalized.includes("manual login")) {
            return {
                status: "failed",
                errorCode: "LOGIN_REQUIRED",
                errorMessage: message
            };
        }
        return {
            status: "failed",
            errorCode: "PROVIDER_RUN_FAILED",
            errorMessage: message
        };
    }
}
